plugins/uartdmx/UartDmxThread.cpp

```diff
@@ -93,46 +105,94 @@ void *UartDmxThread::Run() {
       buffer.Set(m_buffer);
     }
 
-    if (!m_widget->SetBreak(true))
-      goto framesleep;
+    writeDmxData(buffer);
+  }
+  return NULL;
+}
 
-    if (m_granularity == GOOD)
-      usleep(m_breakt);
+/**
+ * Write the DMX data to the actual UART interface.
+ */
+void UartDmxThread::writeDmxData(const DmxBuffer &buffer) {
+  TimeStamp ts1;
+  Clock clock;
 
-    if (!m_widget->SetBreak(false))
-      goto framesleep;
+  // ensures that ts1 has a valid value for the framesleep if break condition cannot be locked/released
+  clock.CurrentMonotonicTime(&ts1);
 
-    if (m_granularity == GOOD)
-      usleep(DMX_MAB);
+  if (!m_widget->SetBreak(true)) {
+    frameSleep(ts1);
+    return;
+  }
 
-    if (!m_widget->Write(buffer))
-      goto framesleep;
+  if (m_granularity == GOOD)
+    usleep(m_breakt);
 
-  framesleep:
-    // Sleep for the remainder of the DMX frame time
-    usleep(m_malft);
+  if (!m_widget->SetBreak(false)) {
+    frameSleep(ts1);
+    return;
   }
-  return NULL;
+
+  // stores the time stamp for the elapsed time calculation during the framesleep
+  clock.CurrentMonotonicTime(&ts1);
+
+  if (m_granularity == GOOD)
+    usleep(DMX_MAB);
+
+  m_widget->Write(buffer);
+  
+  frameSleep(ts1);
 }
 
+/**
+ * Sleeps for the rest of the frame time and tries to recover the granularity if needed. 
+ */
+void UartDmxThread::frameSleep(const TimeStamp &ts1) {
+  TimeStamp ts2, ts3;
+  Clock clock;
+
+  // Sleep for the remainder of the DMX frame time
+  clock.CurrentMonotonicTime(&ts2);
+  TimeInterval elapsed = ts2 - ts1;
+
+  if (m_granularity == GOOD) {
+    while (elapsed.InMilliSeconds() < m_frameTime) {
+      usleep(1000);
+      clock.CurrentMonotonicTime(&ts2);
+      elapsed = ts2 - ts1;
+    }
+  } else {
+    // See if we can drop out of bad mode.
+    usleep(1000);
+    clock.CurrentMonotonicTime(&ts3);
+    TimeInterval interval = ts3 - ts2;
+    if (interval.InMilliSeconds() <= BAD_GRANULARITY_LIMIT) {
+      m_granularity = GOOD;
+      OLA_INFO << "Switching from BAD to GOOD granularity for UART thread";
+    }
+
+    elapsed = ts3 - ts1;
+    while (elapsed.InMilliSeconds() < m_frameTime) {
+      clock.CurrentMonotonicTime(&ts2);
+      elapsed = ts2 - ts1;
+    }
+  }
+}
 
 /**
  * Check the granularity of usleep.
  */
 void UartDmxThread::CheckTimeGranularity() {
   TimeStamp ts1, ts2;
   Clock clock;
-  /** If sleeping for 1ms takes longer than this, don't trust
-   * usleep for this session
-   */
```

Comment:
This comment is really useful to know that 3 isn't the number of times before its bad or something, could you re-attach it to BAD_GRANULARITY_LIMIT please.